python/ray/_raylet.pyx

```diff
@@ -481,6 +482,58 @@ cdef raise_if_dependency_failed(arg):
         raise arg
 
 
+cdef c_bool determine_if_retryable(
+    Exception e,
+    const c_string serialized_retry_exception_allowlist,
+    FunctionDescriptor function_descriptor,
+):
+    """Determine if the provided exception is retryable, according to the
+    (possibly null) serialized predicate function.
+
+    If the serialized predicate function is an empty string or is None once
+    deserialized, the exception is considered retryable and we return True.
+
+    This method can raise exceptions in several cases:
+        - Deserialization fails (TypeError)
+        - Deserialized predicate function is not a function (TypeError)
+        - Predicate function raises an error (RuntimeError)
+        - Predicate function output is not a boolean (TypeError)
+    """
+    if len(serialized_retry_exception_allowlist) == 0:
+        # No exception allowlist specified, default to all retryable.
+        return True
+
+    # Deserialize exception allowlist and check that e is in the allowlist.
+    try:
+        exception_allowlist = ray_pickle.loads(
+            serialized_retry_exception_allowlist,
+        )
+    except TypeError as inner_e:
+        # Exception allowlist deserialization failed.
+        msg = (
+            "Could not deserialize the retry exception allowlist "
+            f"function for task {function_descriptor.repr}. "
+            "Check "
+            "https://docs.ray.io/en/master/ray-core/objects/serialization.html#troubleshooting " # noqa
+            "for more information.")
+        raise TypeError(msg) from inner_e
+
+    if exception_allowlist is None:
+        # No exception allowlist specified, default to all retryable.
+        return True
```

Comment:
Can we do the validation earlier (e.g. in the ray_options_util.py) so here we can assert the deserialization always succeeds? This way we can give better error message before the task runs.

Reply:
We already do validation in `ray_options_util.py`, but it's still possible that an exception allowlist is of the right type (`List[Exception]`), serialization succeeds, and then deserialization on anther machine fails, so we need the above error handling.

As for the exception allowlist being `None`, this is covering the case in which an exception allowlist isn't given (e.g. the user gives `retry_exceptions=True`), where we set the exception allowlist to be `None`: https://github.com/ray-project/ray/pull/25896/files#diff-d70d02f17addb0988bab4dbf8491153b952557d6fa335f9efea0a02fd76d4609R260